### DISCUSSION TITLE
JSON results documentation

### DISCUSSION THREAD

**Discussion by brendonky18** [mandiant/capa#1207] [Q&A] [answered]
Going from version 3 to version 4, the release notes say that the JSON results format was changed, and that these changed were documented. Does anyone have a link to the documentation explaining what changed between v3 and v4?

**Comment by mr-tz** [comment, marked as answer]
Hey, I don't see more detailed documentation on this either, but the major change between v3 and v4 is the change from number-only based addresses, e.g. `"base_address": 4194304,` to more expressive/flexible addresses, e.g.:
```json
"base_address":
{
    "type": "absolute",
    "value": 4194304
},
```

For the update we also switched to using a `pydantic` model. If you're processing the results that should simplify a few things. An example to load the stored results:

```python
with open(capa_results.json, "rb") as f:
    doc = json.loads(f.read().decode("utf-8"))
res = rd.ResultDocument.parse_obj(results)
```

You can then access the result document fields like `res.meta` vs. the `res["meta"]` access before.

Does that help? I'm happy to provide more details for additional questions.

**Reply by brendonky18** [reply]
Yeah that's super helpful, thanks!

Do you know if there are any other valid formats for the `base_address` field besides `absolute` that I would have to account for?

**Reply by mr-tz** [reply]
`base_address` will be `absolute` or the `NO_ADDRESS` address type. All supported address types can be found here for example https://github.com/mandiant/capa/blob/e1735f0a5e2d5561a90119aaebd10c51f036ef3d/capa/features/freeze/__init__.py#L38.

**Reply by brendonky18** [reply]
Looking at what you sent, it seems like the address can also be a relative address. How would I find the base address that's calculated from to get the absolute address again?

**Reply by mr-tz** [reply]
here is where the supported return types are listed:
https://github.com/mandiant/capa/blob/fb99ef56e3e45040349eb96c8ec194dee0d05dc3/capa/features/extractors/base_extractor.py#L93
